### PR TITLE
[WFLY-10879] Deployment is not failing as expected and according to the specification if a @Singleton @Startup @PostConstruct initialization failed

### PR DESCRIPTION
**Pull request by spyrkob**
JIRA: https://issues.jboss.org/browse/WFLY-10879

@dmlloyd could you please take a look at this? The first commit uses StartupCountInterceptor to block access to EJBs unless all @Startup beans are successfully initiated. The second uses similar approach for web components

### CODE REVIEW

**Comment by spyrkob** [comment]
@dmlloyd did you have a chance to look into this?

File: ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java
Comment (dmlloyd) [inline ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java]: These constructs should always be using System.nanoTime(), not currentTimeMillis().

**Review by dmlloyd — CHANGES_REQUESTED** [review]
Otherwise this seems OK.

File: ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java
Comment (spyrkob) [reply ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java]: Thanks, updated to use nanoTime

File: undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java
Comment (dmlloyd) [inline undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java]: Undertow handlers should never block.  This can cause the server to hang.  Instead the requests have to be queued somehow and replayed once the server finishes starting.  Also I'm not really sure that a 404 is the right response if the server takes too long to start.

**Review by dmlloyd — DISMISSED** [review]
Sorry, there's one more problem.  I didn't notice it because it used to be that tests always ended up at the bottom of the PR, but I guess they don't anymore.

File: undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java
Comment (spyrkob) [reply undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java]: Yeah I wasn't sure about that part. I'll take a look at it tomorrow and try to come up with a better solution, thanks.

**Comment by dmlloyd** [comment]
This is now too complex to review in a short amount of time; could you please summarize the operation of the code itself?

**Comment by spyrkob** [comment]
1. `StartupCountdown` and `StartupCountdownInterceptor` is changed to invoke either:
   1. onSuccess callback - when all `@Startup` components finished initialization without exceptions, or
   2. onFailure callback - when any of components fails to initialize
2. `ComponentStartupCountdownHandler` is added to Undertow handlers stack and uses `SuspendRequestQueue` to suspend requests based on `StartupCountdown` callbacks: 
   1. if neither `onSuccess` or `onFailure` were called (ie. components are still starting) - will queue up requests in memory,
   2. if `StartupCountdown` invoked `onFailure` callback - will resume all queued up requests, but instead of processing them as normal, they will be failed. Same will apply to any further request for this deployment,
    3. if `StartupCountdown` invoked `onSuccess` callback - will resume all queued up requests dispatching them to to their remaining handlers. Any further request for this deployment will skip the queue and will be passed to next handler in chain
3. `ModuleDeployment` will prevent the EJB components from being started if `StartupCallback` has not emitted onSuccess callback.

**Review by bstansberry — COMMENTED** [review]
The ComponentStartupCountdownHandler part seems redundant. We start up gracefully by default, and if people want to redeploy an app without reloading/restarting the server they can and should suspend it first. IOW having in-flight requests going while an app is starting is avoidable and not recommended, and we shouldn't be adding complex code to handle people doing things we don't recommend.

**Comment by spyrkob** [comment]
sorry for delay on this.
I rebased and removed request queuing. Since there was no need any more for failed/succeeded callbacks in StartupCountdown I restored it to original form.

If this looks OK I'll squash the commits.

**Comment by darranl** [comment]
Retest this please

**Comment by darranl** [comment]
@tadamski / @chengfang Could you please add a review to this PR - it is presently the oldest PR in the queue being over a year old so we either need to take this to a conclusion or close the PR.

**Review by tadamski — APPROVED** [review]
This fix seems good to me.

**Review by chengfang — APPROVED** [review]

**Comment by spyrkob** [comment]
Rebased the PR. Hopefully that resolves the deps-changed label - there are no dependency changes in the PR

**Comment by darranl** [comment]
Just kicked off CI again, we have approvals so if we get a clean run we should get this in.

**Comment by spyrkob** [comment]
I don't know if those test failures are related - I wasn't able to reproduce them in my environment.
Should I rebase the PR to get rid of the deps-changed label?

**Comment by spyrkob** [comment]
retest this please

**Comment by spyrkob** [comment]
@darranl can I do anything else to get this merged?

**Comment by spyrkob** [comment]
There's one more case that the PR doesn't cover - when the EAR specifies initialize-in-order in application.xml.

Added a commit fixing this problem. @chengfang, @dmlloyd, @tadamski could you please review again? 
@chengfang with the change this should address WFLY-13515 as well.

**Comment by chengfang** [comment]
From July 17, 2020 test runs, Linux JDK11 and Linux JDK8 both failed with 12 test failures in `mixed-domain/src/test/java/org/jboss/as/test/integration/domain/mixed/eap700/MixedDomainDeployment700TestCase`, which seems related to the code change here. Windows JDK11 failed with execution timeout (no specific test failures shown), and I'll try to restart Windows JDK11 job.

**Comment by spyrkob** [comment]
@chengfang I can't reproduce this failures locally - do you know if this is an env issue?

**Comment by darranl** [comment]
The mixed domain failures are environmental, they are going to be temporarily disabled under this PR https://github.com/wildfly/wildfly/pull/13418

**Comment by darranl** [comment]
Re the merge, Jeff is tagging WildFly Core today but will try and have another pass through the PR queue either the end of this week or early next and get approved PRs merged.

**Comment by chengfang** [comment]
Looks good.

**Comment by spyrkob** [comment]
@bstansberry, @dmlloyd I think your comments should be addressed now. Could you approve this if it looks good? There's a customer case related to this, so it'd be great to get this merged :)

File: ejb3/src/main/java/org/jboss/as/ejb3/component/singleton/StartupCountDownInterceptor.java
Comment (spyrkob) [reply ejb3/src/main/java/org/jboss/as/ejb3/component/singleton/StartupCountDownInterceptor.java]: removed, thanks

**Comment by spyrkob** [comment]
@dmlloyd I created PR in Core to add a new constant: https://github.com/wildfly/wildfly-core/pull/4300. The tests will be failing until it is merged though

File: undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java
Comment (bstansberry) [inline undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java:40]: This should be final.

Comment (spyrkob) [reply undertow/src/main/java/org/wildfly/extension/undertow/deployment/ComponentStartupCountdownHandler.java]: fixed, thanks

**Comment by spyrkob** [comment]
@bstansberry core upgrade with wildfly/wildfly-core#4300 have been applied, I think the label can be removed

**Comment by bstansberry** [comment]
Thanks, @spyrkob ! You have great patience. ;)

**Comment by jonmbake** [comment]
This changed caused a subtle bug for us. I wanted to post here in case anyone else runs into the same issue.

We had a EAR structured with module loading order (defined by application.xml) like:

1. ejb1.jar
  - Contains `StartupBean1`
2. web.war
  - Contains a `@WebListener` whose `#contextInitialized` calls into `StartupBean1`
3. ejb2.jar
  - Contains `StartupBean2`

When deploying the ear, it will indefinitely wait for `StartupBean2` to load on the web listener call to `StartupBean1`. It's easily fixable by changing the application.xml to load ejb2.jar before web.war, but it can be challenging figuring out the issue.

**Pull request merged**
